lib/ts/recipe/webauthn/types.ts

```diff
@@ -0,0 +1,608 @@
+    ): Promise<
+        | {
+              status: "OK";
+              webauthnGeneratedOptionsId: string;
+              rp: {
+                  id: string;
+                  name: string;
+              };
+              user: {
+                  id: string;
+                  name: string; // user email
+                  displayName: string; //user email
+              };
+              challenge: string;
+              timeout: number;
+              excludeCredentials: {
+                  id: string;
+                  type: "public-key";
+                  transports: ("ble" | "hybrid" | "internal" | "nfc" | "usb")[];
+              }[];
+              attestation: "none" | "indirect" | "direct" | "enterprise";
+              pubKeyCredParams: {
+                  // we will default to [-8, -7, -257] as supported algorithms. See https://www.iana.org/assignments/cose/cose.xhtml#algorithms
+                  alg: number;
+                  type: "public-key";
+              }[];
+              authenticatorSelection: {
+                  requireResidentKey: boolean;
+                  residentKey: "required" | "preferred" | "discouraged";
+                  userVerification: "required" | "preferred" | "discouraged";
+              };
+          }
+        | { status: "RECOVER_ACCOUNT_TOKEN_INVALID_ERROR" }
+    >;
+
+    signInOptions(input: {
+        relyingPartyId: string;
+        origin: string;
+        userVerification: "required" | "preferred" | "discouraged" | undefined; // see register options
+        timeout: number | undefined;
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<{
+        status: "OK";
+        webauthnGeneratedOptionsId: string;
+        challenge: string;
+        timeout: number;
+        userVerification: "required" | "preferred" | "discouraged";
+    }>;
+
+    signUp(input: {
+        webauthnGeneratedOptionsId: string;
+        credential: {
+            id: string;
+            rawId: string;
+            response: {
+                clientDataJSON: string;
+                attestationObject: string;
+                transports?: ("ble" | "cable" | "hybrid" | "internal" | "nfc" | "smart-card" | "usb")[];
+                userHandle: string;
+            };
+            authenticatorAttachment: "platform" | "cross-platform";
+            clientExtensionResults: Record<string, unknown>;
+            type: "public-key";
+        };
+        session: SessionContainerInterface | undefined;
+        shouldTryLinkingWithSessionUser: boolean | undefined;
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<
+        | {
+              status: "OK";
+              user: User;
+              recipeUserId: RecipeUserId;
+          }
+        | { status: "EMAIL_ALREADY_EXISTS_ERROR" }
+        // when the attestation is checked and is not valid or other cases in whcih the authenticator is not correct
+        | { status: "INVALID_AUTHENTICATOR_ERROR" }
```

Comment:
is there nothing else we need to tell the user when this error happens? Like why was the authenticator invalid?

Reply:
I was thinking that when this kind of error happens, we inform the user to try using another authenticator: "The authenticator is not supported, please try a different one" I could add some reasons tough, but they might be too technical for the users: "Not supported supported authenticator. Try using an Apple device, Android device or supported FIDO authenticator."